### PR TITLE
bpo-32503: Avoid creating too small frames in pickles.

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
https://bugs.python.org/issue32503

### CODE REVIEW

File: Lib/test/pickletester.py [hunk 1]
```diff
@@ -2152,7 +2158,7 @@ def remove_frames(pickled, keep_frame=None):
 
         frame_size = self.FRAME_SIZE_TARGET
         num_frames = 20
-        obj = [bytes([i]) * frame_size for i in range(num_frames)]
+        obj = {i: bytes([i]) * frame_size for i in range(num_frames)}
```
Comment (pitrou) [inline Lib/test/pickletester.py]: IIUC, small integers don't produce a frame? Can you add a comment?

Comment (pitrou) [reply Lib/test/pickletester.py]: I may misunderstand, actually...

File: Lib/test/pickletester.py [hunk 2]
```diff
@@ -2047,36 +2048,43 @@ def check_frame_opcodes(self, pickled):
...
-                # of bytes up to the current frame.
-                frame_size = pos - last_pos - last_frame_opcode_size
-                self.assertEqual(frame_size, last_arg)
-            last_arg, last_pos = arg, pos
-            last_frame_opcode_size = frame_opcode_size
-        # The last frame's size should be equal to the number of bytes up
-        # to the pickle's end.
-        frame_size = len(pickled) - last_pos - last_frame_opcode_size
-        self.assertEqual(frame_size, last_arg)
+                    self.assertLessEqual(pos, frame_end)
```
Comment (pitrou) [inline Lib/test/pickletester.py]: This is repeated just below.

File: Lib/test/pickletester.py [hunk 3]
```diff
@@ -2047,36 +2048,43 @@ def check_frame_opcodes(self, pickled):
...
+                self.assertLessEqual(pos, frame_end)
+                if pos < frame_end:
+                    self.assertNotIn(op.name, 'FRAME')
+                    if op.name in frameless_opcodes:
+                        self.assertLessEqual(len(arg), self.FRAME_SIZE_TARGET)
```
Comment (pitrou) [inline Lib/test/pickletester.py]: Could you add a comment explaining this?

File: Lib/test/pickletester.py [hunk 4]
```diff
@@ -2047,36 +2048,43 @@ def check_frame_opcodes(self, pickled):
...
-            last_arg, last_pos = arg, pos
-            last_frame_opcode_size = frame_opcode_size
-        # The last frame's size should be equal to the number of bytes up
-        # to the pickle's end.
-        frame_size = len(pickled) - last_pos - last_frame_opcode_size
-        self.assertEqual(frame_size, last_arg)
+                    self.assertLessEqual(pos, frame_end)
+                self.assertLessEqual(pos, frame_end)
+                if pos < frame_end:
+                    self.assertNotIn(op.name, 'FRAME')
```
Comment (pitrou) [inline Lib/test/pickletester.py]: This test looks strange. Perhaps you meant `assertNotEqual(op.name, 'FRAME')`?

File: Lib/test/pickletester.py [hunk 5]
```diff
@@ -2047,36 +2048,43 @@ def check_frame_opcodes(self, pickled):
...
+                    if op.name in frameless_opcodes:
+                        self.assertLessEqual(len(arg), self.FRAME_SIZE_TARGET)
+                else:
+                    frame_end = None
+                    #frameless_start = pos
```
Comment (pitrou) [inline Lib/test/pickletester.py]: Hmm, looks like this should be removed.

File: Lib/test/pickletester.py [hunk 6]
```diff
@@ -2047,36 +2048,43 @@ def check_frame_opcodes(self, pickled):
...
+
+            if frame_end is None:
+                if (op.name == 'FRAME' or
+                    (op.name in frameless_opcodes and
+                     len(arg) > self.FRAME_SIZE_TARGET)):
+                    if frameless_start is not None:
+                        self.assertLess(pos - frameless_start,
```
Comment (pitrou) [inline Lib/test/pickletester.py]: Perhaps comment on this?

File: Lib/test/pickletester.py (hunk 2 above)
Comment (serhiy-storchaka) [reply Lib/test/pickletester.py]: Good point. I tried to write these tests in different forms and some checks are left duplicated.

File: Lib/test/pickletester.py (hunk 4 above)
Comment (serhiy-storchaka) [reply Lib/test/pickletester.py]: Right.

File: Lib/test/pickletester.py (hunk 3 above)
Comment (serhiy-storchaka) [reply Lib/test/pickletester.py]: Done.

File: Lib/test/pickletester.py (hunk 6 above)
Comment (serhiy-storchaka) [reply Lib/test/pickletester.py]: Done.

File: Lib/test/pickletester.py (hunk 5 above)
Comment (serhiy-storchaka) [reply Lib/test/pickletester.py]: Done.

File: Lib/test/pickletester.py (hunk 1 above)
Comment (serhiy-storchaka) [reply Lib/test/pickletester.py]: I have added small objects (dict keys) between large objects.

**Review by pitrou — APPROVED** [review]
Thanks for addressing the comments!

**Comment by serhiy-storchaka** [comment]
Thank you for your review!

**Pull request merged**